### ISSUE TITLE
[ENH] Utility for catching GPUs in exclusive mode

### ISSUE THREAD

**Issue by IAlibay** [choderalab/openmmtools#697]
After discussions with @mikemhenry we think this might need to live here.

We have had issues with folks trying to use a multistatesampler derived class (repex, sams, etc...) and things failing in a non-clean manner beause their GPU setup was set in exclusive mode (tagging @ijpulidos who found this issue at the OMSF workshop).

It would be good to have a utility here that does a call to `nvidia-smi --query-gpu="compute_mode" --format=csv` or similar to get the compute mode and check that we aren't in exclusive mode.

**Comment by ijpulidos** [comment]
Yes, this is something we can do,  detect if the device is in `Exclusive_Process` compute mode and stop early if this is the case, because the subsequent errors are kinda cryptic. 

On the other hand, maybe there could be something else going on here. This issue seems to suggest that we are running more than job in the GPU and, if so, why is it that we need to run more than one job on the device? Just trying to think if this is a symptom of not deleting/removing contexts properly in the device from our part.

**Comment by ijpulidos** [comment]
FYI. When I monitor the usage of the GPU when running simulations I only see one process, as expected.

**Comment by IAlibay** [comment]
> Just trying to think if this is a symptom of not deleting/removing contexts properly in the device from our part.

Yeah so my thought here was that maybe we're keeping GPU contexts for the replicas alive on the CUDA device (even though only one process is running at a given time), so it's being registered as a multi-process execution.

**Comment by mikemhenry** [comment]
I think it is one process with more than one context maybe? I can't work on this right now, but when I get around to it (or someone else) if you have a GPU on a headless box, you can put it in `Exclusive_Process` mode and see what happens.

**Comment by jchodera** [comment]
+1 for adding intelligent error checking to warn the user if they are trying to use thread- or process-exclusive mode with CUDA. 

The overhead with creating CUDA contexts is large enough that we generally cache multiple OpenMM `Context` objects to avoid having to keep creating and destroying them, but many CUDA installations are set up to use thread- or process-exclusive mode by default, which won't allow more than one CUDA context without triggering an exception.

@mikemhenry : Would you be able to tackle this?

**Comment by mikemhenry** [comment]
Yes!

**Comment by ijpulidos** [comment]
We probably can tackle https://github.com/choderalab/openmmtools/issues/693 in the same set of changes

**Comment by IAlibay** [comment]
@ijpulidos could looking at CUDA_VISIBLE_DEVICES work? If multiple devices are present and also defined in the env var or the env var is not defined then I think the assumption of only needing to default to 0 should be fine? (iirc OpenMM defaults to visible device 0 right?)

**Comment by ijpulidos** [comment]
> @ijpulidos could looking at CUDA_VISIBLE_DEVICES work?

Maybe, but it gets kinda tricky. `CUDA_VISIBLE_DEVICES` uses the API order, which orders devices by compute capabilities, whereas `nvidia-smi` orders devices by PCI ID (which depends on the OS). So we cannot guarantee correspondence there.

**Comment by ijpulidos** [comment]
We could also use the nvidia UUID (not just the index) to work around this limitation.

We can try-catch the `openmm.OpenMMException: No compatible CUDA device is available` and try debugging that once it's caught.

**Comment by mikemhenry** [comment]
```
"serial" or "gpu_serial"
This number matches the serial number physically printed on each board. It is a globally unique immutable alphanumeric value.

"uuid" or "gpu_uuid"
This value is the globally unique immutable alphanumeric identifier of the GPU. It does not correspond to any physical label on the board.
```

Do we want to log the `uuid` or the `serial` number?

**Comment by mikemhenry** [comment]
Ah if the GPU is integrated, there is no serial number printed and comes back as NA, so I will use the `uuid`
